File: 155. Min Stack/solution.py
import heapq

class MinStackArray:
    def __init__(self):
        self.stack = [] 
        self.mini = []

    def push(self, val: int) -> None:
        if not self.mini or val <= self.mini[-1]: 
            self.mini.append(val)
        self.stack.append(val)

    def pop(self) -> None:
        val = self.stack.pop()
        if self.mini and val == self.mini[-1]:
            self.mini.pop()
    
    def top(self) -> int:
        return self.stack[-1]
        
    def getMin(self) -> int:
        return self.mini[-1]

File: 155. Min Stack/solution.cpp
#include <algorithm>
#include <vector>


using namespace std;

class MinStack {
public:
    vector<int> stack; 
    vector<int> mini;

    MinStack() {
    }
    
    void push(int val) {
        stack.push_back(val);
        if (mini.empty()) {
            mini.push_back(val);
        } else {
            mini.push_back(min(val, mini.back()));
        }
    }
    
    void pop() {
      if (stack.back() == mini.back()) {
        mini.pop_back();
      }
      stack.pop_back();
    }
    
    int top() {
      return stack.back();
    }
    
    int getMin() {
      return mini.back();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
